assets/js/util/report-validation.js

```diff
@@ -0,0 +1,134 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Validates data that can be either string or object of the certain type, or array of them.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|string[]|Object|Object[]} data The data to check.
+ * @param {Function} objectVerificationCallback The callback to verify an object.
+ * @return {boolean} TRUE if data is valid, otherwise FALSE.
+ */
+function isValidStringsOrObjects( data, objectVerificationCallback ) {
+	if ( typeof data === 'string' ) {
+		return true;
+	}
+
+	if ( typeof data === 'object' && objectVerificationCallback.call( null, data ) ) {
+		return true;
+	}
+
+	if ( Array.isArray( data ) ) {
+		for ( let i = 0; i < data.length; i++ ) {
+			const type = typeof data[ i ];
+			if ( type !== 'string' && ( type !== 'object' || ! objectVerificationCallback.call( null, data[ i ] ) ) ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * Verifies that provided metrics match allowed values. Metrics can be a string,
+ * an array of string, an array of objects or mix of strings and objects. Objects
+ * must have "expression" and "alias" properties in order to be considered as valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|string[]|Object|Object[]} metrics The metrics to check.
+ * @return {boolean} TRUE if metrics are valid, otherwise FALSE.
+ */
+export function isValidMetrics( metrics ) {
+	return isValidStringsOrObjects( metrics, ( metric ) => {
+		const validExpression = metric.hasOwnProperty( 'expression' ) && typeof metric.expression === 'string';
+		const validAlias = metric.hasOwnProperty( 'alias' ) && typeof metric.alias === 'string';
+		return validExpression && validAlias;
+	} );
+}
+
+/**
+ * Verifies provided dimensions to make sure it matches allowed values. It can be a string,
+ * array of strings, an object with "name" field, array of such objects or an array of strings
+ * and objects.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string|string[]|Object|Object[]} dimensions The dimensions to check.
+ * @return {boolean} TRUE if dimensions are valid, otherwise FALSE.
+ */
+export function isValidDimensions( dimensions ) {
+	return isValidStringsOrObjects( dimensions, ( dimension ) => {
+		return dimension.hasOwnProperty( 'name' ) && typeof dimension.name === 'string';
+	} );
+}
+
+/**
+ * Verifies that either date range or start and end dates are valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} dateRange The date range to check.
+ * @param {string} startDate The start date to check.
+ * @param {string} endDate The end date to check.
+ * @return {boolean} TRUE if either date range or start/end dates are valid, otherwise FALSE.
+ */
+export function isValidDateRange( dateRange, startDate, endDate ) {
```

Comment:
Because not all arguments are required, we shouldn't use positional arguments here. A function signature like: 
```suggestion
export function isValidDateRange( { dateRange, startDate, endDate } = {} ) {
```

Would allow you to check `startDate` and `endDate` without needing to pass a bogus first argument to the function.

Reply:
Updated.